File: repositories/open-agentic-platform/product/packages/prompt-assembly/src/assembler.ts
// SPDX-License-Identifier: AGPL-3.0-or-later
// Feature: 070-prompt-assembly-cache

import type {
  AssembledPrompt,
  AssemblerOptions,
  AssemblyContext,
  AssemblyMetadata,
  PromptSection,
  SectionMetadata,
} from "./types.js";

/**
 * Machine-readable boundary marker between static and dynamic prompt regions.
 * The API client uses this to place the cache breakpoint.
 */
export const CACHE_BOUNDARY_MARKER =
  "\n═══ CACHE BOUNDARY ═══\n";

const DEFAULT_TOTAL_BUDGET = 102_400; // 100 KB

/**
 * Truncate content to fit within a byte budget.
 * If truncated, appends a notice so the model knows content was cut.
 */
export function truncateToBytes(content: string, maxBytes: number): string {
  const encoder = new TextEncoder();
  const bytes = encoder.encode(content);
  if (bytes.length <= maxBytes) return content;

  const notice = "\n[... truncated to fit budget ...]";
  const noticeBytes = encoder.encode(notice).length;
  const available = Math.max(0, maxBytes - noticeBytes);

  // Truncate at byte level, then decode back (handles multi-byte safely)
  const truncated = new TextDecoder().decode(bytes.slice(0, available));
  return truncated + notice;
}

/**
 * PromptAssembler composes registered sections into a two-region system prompt
 * with a cache boundary between static and dynamic content.
 *
 * Static and per_session sections are cached after first assembly.
 * Dynamic sections are rebuilt every turn.
 *
 * Sections are ordered by priority (descending). When the total budget is
 * exceeded, remaining sections are dropped with a truncation notice.
 *
 * FR-001: Two regions separated by cache boundary
 * FR-002: Sections registered with name, contentFn, cacheLifetime, priority, maxBytes
 * FR-005: Budget enforcement with truncation from lowest-priority
 * FR-007: Runtime registration via registerSection()
 * FR-008: Structured metadata emission
 * NF-001: Assembly completes in <10ms for 30 sections (excluding content generation)
 * NF-002: Static content is deterministic for cache effectiveness
 */
export class PromptAssembler {
  private sections: PromptSection[] = [];
  private readonly cache = new Map<string, string>();
  private readonly totalBudget: number;

  constructor(options: AssemblerOptions = {}) {
    this.totalBudget = options.totalBudget ?? DEFAULT_TOTAL_BUDGET;
  }

  /**
   * Register a new prompt section. (FR-007)
   * Sections are re-sorted by priority descending on each registration.
   */
  registerSection(section: PromptSection): void {
    this.sections.push(section);
    this.sections.sort((a, b) => b.priority - a.priority);
  }

  /**
   * Remove a section by name.
   */
  removeSection(name: string): boolean {
    const before = this.sections.length;
    this.sections = this.sections.filter((s) => s.name !== name);
    this.cache.delete(name);
    return this.sections.length < before;
  }

  /**
   * Invalidate the cache for a specific section (e.g., when CLAUDE.md changes).
   * Pass no arguments to invalidate the entire cache.
   */
  invalidateCache(name?: string): void {
    if (name) {
      this.cache.delete(name);
    } else {
      this.cache.clear();
    }
  }

  /**
   * Return a snapshot of registered section names in priority order.
   */
  get sectionNames(): readonly string[] {
    return this.sections.map((s) => s.name);
  }

  /**
   * Assemble the full system prompt. (FR-001, FR-005, FR-008)
   *
   * 1. Iterate sections in priority order (descending).
   * 2. Resolve content — from cache for static/per_session, fresh for dynamic.
   * 3. Truncate per-section if over maxBytes.
   * 4. Stop adding sections when total budget exceeded.
   * 5. Separate static and dynamic parts with the cache boundary marker.
   */
  assemble(ctx: AssemblyContext): AssembledPrompt {
    const staticParts: string[] = [];
    const dynamicParts: string[] = [];
    const sectionMetas: SectionMetadata[] = [];
    const truncatedSections: string[] = [];
    let totalSize = 0;
    let budgetExceeded = false;

    for (const section of this.sections) {
      // Resolve content, using cache for static/per_session
      let cacheHit = false;
      let content: string;

      if (
        section.cacheLifetime === "static" ||
        section.cacheLifetime === "per_session"
      ) {
        const cached = this.cache.get(section.name);
        if (cached !== undefined) {
          content = cached;
          cacheHit = true;
        } else {
          content = section.contentFn(ctx);
          this.cache.set(section.name, content);
        }
      } else {
        content = section.contentFn(ctx);
      }

      const originalBytes = new TextEncoder().encode(content).length;

      // Per-section truncation
      const truncated = truncateToBytes(content, section.maxBytes);
      const finalBytes = new TextEncoder().encode(truncated).length;
      const wasTruncated = finalBytes < originalBytes;

      // Check total budget
      if (totalSize + finalBytes > this.totalBudget) {
        budgetExceeded = true;
        truncatedSections.push(section.name);
        // Record metadata for dropped section
        sectionMetas.push({
          name: section.name,
          cacheLifetime: section.cacheLifetime,
          priority: section.priority,
          originalBytes,
          finalBytes: 0,
          truncated: true,
          cacheHit,
        });
        continue;
      }

      totalSize += finalBytes;

      if (wasTruncated) {
        truncatedSections.push(section.name);
      }

      sectionMetas.push({
        name: section.name,
        cacheLifetime: section.cacheLifetime,
        priority: section.priority,
        originalBytes,
        finalBytes,
        truncated: wasTruncated,
        cacheHit,
      });

      // Route to static or dynamic bucket
      if (
        section.cacheLifetime === "static" ||
        section.cacheLifetime === "per_session"
      ) {
        staticParts.push(truncated);
      } else {
        dynamicParts.push(truncated);
      }
    }

    const metadata: AssemblyMetadata = {
      totalBytes: totalSize,
      sectionCount: staticParts.length + dynamicParts.length,
      sections: sectionMetas,
      truncatedSections,
      budgetExceeded,
    };

    return {
      staticPrefix: staticParts.join("\n"),
      cacheBoundary: CACHE_BOUNDARY_MARKER,
      dynamicSuffix: dynamicParts.join("\n"),
      metadata,
    };
  }

  /**
   * Convenience: assemble and return the full prompt as a single string.
   */
  assembleString(ctx: AssemblyContext): string {
    const result = this.assemble(ctx);
    return (
      result.staticPrefix + result.cacheBoundary + result.dynamicSuffix
    );
  }
}

File: repositories/open-agentic-platform/crates/agent/src/prompt/assembler.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (C) 2026 Bartek Kus
// Feature: 070-PROMPT_ASSEMBLY_CACHE

//! Section registry and prompt assembly pipeline.
//!
//! The assembler composes registered [`PromptSection`]s into an [`AssembledPrompt`]
//! with a static prefix (cacheable across turns) separated from a dynamic suffix
//! by [`CACHE_BOUNDARY_MARKER`].

use std::collections::HashMap;

/// Machine-readable separator between static and dynamic prompt content (FR-001).
/// The API client uses this to place the cache breakpoint.
pub const CACHE_BOUNDARY_MARKER: &str = "═══ CACHE BOUNDARY ═══";

/// Default total prompt budget in bytes (FR-005).
const DEFAULT_TOTAL_BUDGET: usize = 100 * 1024; // 100 KB

/// Default per-section budget when none is specified.
const DEFAULT_SECTION_BUDGET: usize = 32 * 1024; // 32 KB

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/// Cache lifetime for a prompt section (FR-002).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CacheLifetime {
    /// Assembled once, reused across turns.
    Static,
    /// Assembled once per session, cached within the session.
    PerSession,
    /// Rebuilt every turn.
    Dynamic,
}

/// Shared context threaded through content functions during assembly.
#[derive(Debug, Default)]
pub struct AssemblyContext {
    /// Key-value bag for arbitrary context (workflow state, session id, etc.).
    pub values: HashMap<String, String>,
}

/// A registered prompt section (FR-002).
pub struct PromptSection {
    /// Unique section name.
    pub name: String,
    /// Returns the section content for the current context.
    pub content_fn: Box<dyn Fn(&AssemblyContext) -> String + Send + Sync>,
    /// Determines caching behavior.
    pub cache_lifetime: CacheLifetime,
    /// Higher priority = earlier in the prompt.
    pub priority: u32,
    /// Per-section byte budget.
    pub max_bytes: usize,
}

/// Per-section metadata emitted after assembly (FR-008).
#[derive(Debug, Clone)]
pub struct SectionSummary {
    pub name: String,
    pub bytes: usize,
    pub cache_hit: bool,
    pub truncated: bool,
}

/// Metadata about the assembled prompt (FR-008).
#[derive(Debug, Clone)]
pub struct AssemblyMetadata {
    pub total_bytes: usize,
    pub section_count: usize,
    pub sections: Vec<SectionSummary>,
    pub truncated_sections: Vec<String>,
    pub budget_exceeded: bool,
}

/// The fully assembled prompt produced by [`PromptAssembler::assemble`] (FR-001).
#[derive(Debug, Clone)]
pub struct AssembledPrompt {
    /// Content that should be cached across turns.
    pub static_prefix: String,
    /// The cache boundary marker.
    pub cache_boundary: String,
    /// Content rebuilt each turn.
    pub dynamic_suffix: String,
    /// Assembly observability data.
    pub metadata: AssemblyMetadata,
}

impl AssembledPrompt {
    /// Render the full prompt as a single string.
    pub fn render(&self) -> String {
        if self.dynamic_suffix.is_empty() {
            self.static_prefix.clone()
        } else {
            format!(
                "{}\n{}\n{}",
                self.static_prefix, self.cache_boundary, self.dynamic_suffix
            )
        }
    }
}

// ---------------------------------------------------------------------------
// Assembler
// ---------------------------------------------------------------------------

/// Prompt assembler with section registry, caching, and budget enforcement (FR-001 – FR-008).
pub struct PromptAssembler {
    sections: Vec<PromptSection>,
    total_budget: usize,
    cache: HashMap<String, String>,
}

impl PromptAssembler {
    /// Create an assembler with the default 100 KB budget.
    pub fn new() -> Self {
        Self {
            sections: Vec::new(),
            total_budget: DEFAULT_TOTAL_BUDGET,
            cache: HashMap::new(),
        }
    }

    /// Create an assembler with a custom total budget.
    pub fn with_budget(total_budget: usize) -> Self {
        Self {
            sections: Vec::new(),
            total_budget,
            cache: HashMap::new(),
        }
    }

    /// Register a new prompt section at runtime (FR-007).
    pub fn register_section(&mut self, section: PromptSection) {
        self.sections.push(section);
        // Re-sort so highest priority comes first.
        self.sections
            .sort_by_key(|s| std::cmp::Reverse(s.priority));
    }

    /// Invalidate the cache entry for a named section (R-001 mitigation).
    pub fn invalidate(&mut self, name: &str) {
        self.cache.remove(name);
    }

    /// Clear all cached content.
    pub fn invalidate_all(&mut self) {
        self.cache.clear();
    }

    /// Number of registered sections.
    pub fn section_count(&self) -> usize {
        self.sections.len()
    }

    /// Assemble the full prompt from registered sections (FR-001, FR-005, FR-008).
    pub fn assemble(&mut self, ctx: &AssemblyContext) -> AssembledPrompt {
        let mut static_parts: Vec<String> = Vec::new();
        let mut dynamic_parts: Vec<String> = Vec::new();
        let mut total_size: usize = 0;
        let mut section_summaries: Vec<SectionSummary> = Vec::new();
        let mut truncated_sections: Vec<String> = Vec::new();
        let mut budget_exceeded = false;

        for section in &self.sections {
            // Resolve content — use cache for static/per-session (FR-002).
            let cache_hit;
            let content = match section.cache_lifetime {
                CacheLifetime::Static | CacheLifetime::PerSession => {
                    if let Some(cached) = self.cache.get(&section.name) {
                        cache_hit = true;
                        cached.clone()
                    } else {
                        cache_hit = false;
                        let generated = (section.content_fn)(ctx);
                        self.cache.insert(section.name.clone(), generated.clone());
                        generated
                    }
                }
                CacheLifetime::Dynamic => {
                    cache_hit = false;
                    (section.content_fn)(ctx)
                }
            };

            // Per-section truncation (FR-005).
            let budget = if section.max_bytes > 0 {
                section.max_bytes
            } else {
                DEFAULT_SECTION_BUDGET
            };
            let was_truncated = content.len() > budget;
            let truncated = if was_truncated {
                truncate_to_budget(&content, budget)
            } else {
                content
            };

            // Total budget check (FR-005).
            if total_size + truncated.len() > self.total_budget {
                truncated_sections.push(section.name.clone());
                budget_exceeded = true;
                // Continue checking remaining sections — they're all dropped.
                for remaining in self.sections.iter().skip(section_summaries.len() + 1) {
                    truncated_sections.push(remaining.name.clone());
                }
                break;
            }

            total_size += truncated.len();

            section_summaries.push(SectionSummary {
                name: section.name.clone(),
                bytes: truncated.len(),
                cache_hit,
                truncated: was_truncated,
            });

            match section.cache_lifetime {
                CacheLifetime::Static | CacheLifetime::PerSession => {
                    static_parts.push(truncated);
                }
                CacheLifetime::Dynamic => {
                    dynamic_parts.push(truncated);
                }
            }
        }

        let section_count = section_summaries.len();

        AssembledPrompt {
            static_prefix: static_parts.join("\n"),
            cache_boundary: CACHE_BOUNDARY_MARKER.to_string(),
            dynamic_suffix: dynamic_parts.join("\n"),
            metadata: AssemblyMetadata {
                total_bytes: total_size,
                section_count,
                sections: section_summaries,
                truncated_sections,
                budget_exceeded,
            },
        }
    }
}

impl Default for PromptAssembler {
    fn default() -> Self {
        Self::new()
    }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/// Truncate content to `max_bytes`, breaking at a UTF-8 char boundary,
/// and appending a truncation notice.
fn truncate_to_budget(content: &str, max_bytes: usize) -> String {
    const NOTICE: &str = "\n[... truncated]";
    if content.len() <= max_bytes {
        return content.to_string();
    }
    let usable = max_bytes.saturating_sub(NOTICE.len());
    // Walk back to a char boundary.
    let mut end = usable;
    while end > 0 && !content.is_char_boundary(end) {
        end -= 1;
    }
    let mut out = content[..end].to_string();
    out.push_str(NOTICE);
    out
}
